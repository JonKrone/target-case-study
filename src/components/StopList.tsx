import React from 'react'

export const StopList: React.FC<{}> = () => {
  return (
    <section
      style={{
        backgroundColor: 'cadetblue',
        flexGrow: 5,
      }}
    >
      {false ? (
        <>No route selected!</>
      ) : (
        <>
          <h1 style={{ textAlign: 'center' }}>Stops</h1>
          <ul>
            <li>Stop 1</li>
          </ul>
        </>
      )}
    </section>
  )
}

export default StopList
