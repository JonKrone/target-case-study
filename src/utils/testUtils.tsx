import React, { ReactElement } from 'react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { render } from '@testing-library/react'

export const renderWithWrapper = (
  ui: ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: any } = {}
) => {
  const Wrapper: React.FC<{}> = ({ children }) => (
    <Router history={history}>{children}</Router>
  )

  return {
    ...render(ui, { wrapper: Wrapper }),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  }
}
