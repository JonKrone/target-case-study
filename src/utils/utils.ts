export const capitalize = (word: string) =>
  word[0].toLocaleUpperCase() + word.slice(1).toLocaleLowerCase()

export const cardinalNumMap: Record<string, string> = {
  1: 'SOUTHBOUND',
  2: 'EASTBOUND',
  3: 'WESTBOUND',
  4: 'NORTHBOUND',
  SOUTHBOUND: '1',
  EASTBOUND: '2',
  WESTBOUND: '3',
  NORTHBOUND: '4',
}
