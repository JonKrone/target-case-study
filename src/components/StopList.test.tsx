import React from 'react'
import { renderWithWrapper } from '../utils/testUtils'

import StopList from './StopList'
import { Route } from 'react-router-dom'

jest.mock('use-http', () => ({
  __esModule: true,
  default: jest.fn(() => mockUseFetchResult),
}))

const mockUseFetchResult = {
  get: jest.fn(() => {}),
  data: [
    { Text: 'Stop 1', Value: '1' },
    { Text: 'Stop 2', Value: '2' },
  ],
}

describe('StopItem', () => {
  it('renders with no selection', () => {
    const { asFragment, getByText } = renderWithWrapper(<StopList />)

    expect(getByText(/select a route/)).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('receives the Direction and Stop to fetch from URL params', () => {
    const { asFragment } = renderWithWrapper(<StopList />, { route: '/81/1' })
  })

  it('renders with selection', () => {
    // I'm unsure whether it's better to mock react-router's `useParams` here,
    // or to wrap the StopList in a Route like this.
    const { asFragment, getByText, history } = renderWithWrapper(
      <Route path="/:route/:direction">
        <StopList />
      </Route>,
      {
        route: '/81/1',
      }
    )

    expect(getByText(/Stops/)).toBeInTheDocument()
    expect(getByText(/Southbound/)).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('fetches Stops', () => {
    const { asFragment, getByText } = renderWithWrapper(
      <Route path="/:route/:direction">
        <StopList />
      </Route>,
      {
        route: '/81/1',
      }
    )

    expect(mockUseFetchResult.get).toHaveBeenCalledWith('/Stops/81/1')
  })
})
