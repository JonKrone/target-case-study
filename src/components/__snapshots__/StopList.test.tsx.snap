// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopItem renders with no selection 1`] = `
<DocumentFragment>
  <section
    style="flex: 1; overflow-y: auto; border-bottom-right-radius: 5px;"
  >
    <h1
      style="text-align: center;"
    >
      Please select a route
    </h1>
  </section>
</DocumentFragment>
`;

exports[`StopItem renders with selection 1`] = `
<DocumentFragment>
  <section
    style="flex: 1; overflow-y: auto; border-bottom-right-radius: 5px;"
  >
    <h1
      style="text-align: center;"
    >
      Southbound Stops
    </h1>
    <ol
      style="padding-left: 3rem;"
    >
      <li
        style="padding-top: 0.5rem;"
      >
        Stop 1
      </li>
      <li
        style="padding-top: 0.5rem;"
      >
        Stop 2
      </li>
    </ol>
  </section>
</DocumentFragment>
`;
